### DISCUSSION TITLE
Regular expression no longer matching when used in a syntax command

### DISCUSSION THREAD

**Discussion by adrianconstantin-leroy** [vim/vim#14909] [Q&A]
### Steps to reproduce

Open new Vim window and paste this (a pdoc (php) man page header):

```
ARRAY(3)                                                   1                                                   ARRAY(3)

array - Create an array

SYNOPSIS
       array array ([mixed  $...])

DESCRIPTION
        Creates an array. Read the section on the array type for more information on what an array is.

PARAMETERS
              o $...
                -  Syntax  "index => values", separated by commas, define index and values. index may be of type string
```

Search for the following pattern, that matches the first non-empty lines in the synopsis header

```Vim
:set hls
/\v^SYNOPSIS\r?\n\zs.*\ze\r?\n\r?\n
```

Notice how the first line of the synopsis header gets highlighted.

Now use the same regular expression in a `:syntax match` command:

```Vim
:set nohls
:syntax clear
:syntax match pdocSynopsisHeader /\v^SYNOPSIS\r?\n\zs.*\ze\r?\n\r?\n/
:highlight link pdocSynopsisHeader Special
```

Notice how the syntax does not change, and there is no match for the `:syntax match` command, despite the same regular expression is used as before.

You can double-check by moving the input cursor to the line in question (line 6) and calling synstack() function:

```Vim
:normal 6G
:echo synstack(line('.'), col('.'))->map( { _, val -> synIDattr(val, "name") })
:syntax list pdocSynopsisHeader
```

What am I missing ? Some multi-line flag for the regexp ... ?

### Expected behaviour

The regexp should match in the same way when used for a search (with */* ) or when used in the :syntax command.
Or otherwise any differences should be documented.

### Version of Vim

9.1 (patches 1-445)

### Environment

Windows 10 GUI (gvim.exe)

**Comment by chrisbra** [comment]
This seems more like a usage question and not necessarily a bug (at least not yet).

I see several issues here:
- first of all, your use of `\r?\n` is not required, you can always use `\n` to match a linebreak, independently of what the fileformat option is set.
- I seem to remember some issues with `\zs` and zero-width matches for syntax matches (not sure if this is documented). The following change works however: `syn match Special /\(^SYNOPSIS\n\)\@<=.*$/`. I think the difference is where the syntax engine is trying to start the match.
- finally, I think using multi-line patterns as a `:syn-match` is not necessarily best-practice for syntax scripts. Better it is to define a region start and end patterns and then explicitly try to match within that region. So try this one instead (which is also more simpler to read:
```
syn region SYNOPSIS start="^SYNOPSIS" end="^$" contains=phpDescription
syn match phpDescription /^\s\+.*$/ contained
hi def link phpDescription Special
```

**Comment by adrianconstantin-leroy** [comment]
I was able to write a syntax region, and using `@<=` instead of `\zs` worked.

I tried this before, my mistake was to use `\v` for very magic in my regular expression, and then still escape the at sign like `\@<=`    :(

I also tried `ms=e` and `rs=e`, and they would not work:

- using `ms=e` shows no match
- using `rs=e` matches, but the pattern offset has no effect

```Vim
syntax clear | syntax match Statement /^\a.*$/ | syntax region Special start="\v(^(CLASS )?SYNOPSIS$)"ms=e end=/\v\ze\n@<=\a/ keepend
syntax clear | syntax match Statement /^\a.*$/ | syntax region Special start="\v(^(CLASS )?SYNOPSIS$)"rs=e end=/\v\ze\n@<=\a/ keepend
```

Plus the expressions behave differently (worse) when I use `\n` instead of `$` ... is that expected ?

It's a good thing at least one anchor method works (`@<=`), out of the possible 3 (`\zs`, `@<=`, `ms=e`, `rs=e`).

**Comment by nv-quan** [comment]
Why is this an announcement?

**Reply by k-takata** [reply]
Changed to Q&A.

**Reply by chrisbra** [reply]
I have no idea, why it moves to announcements, if I transfer an issue to discussions 🤷

**Comment by dkearns** [comment]
See `:help syn-multi-line`

**Comment by zzzyxwvut** [comment]
Moving `\zs` in the first line should bring about the desired  
effect:

```vim
syntax match pdocSynopsisHeader /\v^SYNOPSIS\r?\zs\n.*\ze\r?\n\r?\n/
```

But take heed of the advice provided above by @chrisbra.

**Comment by adrianconstantin-leroy** [comment]
Thank you guys !

The `\zs` behaviour is still surprising 'though.

After reading `:help syn-multi-line`. the restriction to "stay on the first line" should apply for a `:syntax region` start match. But now it also applies to a `:syntax match` (with a multi-line regexp).

It would help if this limitation was documented under `:help /\zs`, because there the text says "Matches at any position..." at the start of the help text. Maybe also link to the alternative `@<=`.

And the restriction is still a bit under-specified with respect to the `$` anchor. The way I see it, `$` is a zero-width match (anchor) right before the literal LF character (or CR LF), or before the end of file / end of string.

So if I place my `\zs` right after `$`, and `\zs` is also zero-width, I would expect my new anchor to still be right in front of the literal LF character, so still on the first line ... but the help text does not explain or confirm this.

The same is true for pattern offset (match start `ms=e`  in my last example). But the behavior, with my previous example above, does not confirm my logic.

About pattern offset, what exactly happens to the text between match start (`ms`) and region start `rs` ? And the text between region start `rs` and highlighting start  `hs` ? 

And lastly let me say how the pattern offset, or `\zs` position, usually depends on the language syntax (that I want to highlight), and not on the developer choice. And many languages ignore line breaks. As a Vim script developer, all I can do is try to work around this restriction to "stay on the first line" ...  Any chance this restriction can be fixed in Vim ?
